c/cert/src/rules/DCL31-C/DeclareIdentifiersBeforeUsingThem.ql

```diff
@@ -0,0 +1,22 @@
+/**
+ * @id c/cert/declare-identifiers-before-using-them
+ * @name DCL31-C: Declare identifiers before using them
+ * @description Omission of type specifiers may not be supported by some compilers.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/cert/id/dcl31-c
+ *       correctness
+ *       readability
+ *       external/cert/obligation/rule
+ */
+
+import cpp
+import codingstandards.c.cert
+
+from Declaration d
+where
+  not isExcluded(d, Declarations1Package::declareIdentifiersBeforeUsingThemQuery()) and
+  exists(Specifier s | s = d.getASpecifier() and s.toString() = "implicit_int") and
+  exists(d.getFile().getRelativePath())
```

Comment:
"This line isn't necessary because (a) when running the query on real databases the interpretation phase will exclude results without a relative path and (b) when running in tests, the isExcluded predicate will exclude results without a relative path." 

I thought so too! but for openpilot and rule DCL37-C that predicate makes a difference in the results, without it it gets results in files like: `usr/include/x86_64-linux-gnu/bitss/stdlib-bsearch.h` that I thought it would make more sense to omit than report.

similarly, if I add `#include "stdio.h"` to the test file for that rule and run `codeql test run --show-extractor-output --search-path .  c/common/test/rules/declaredareservedidentifier/DeclaredAReservedIdentifier.ql` I start to get some results in that file `+| ../../includes/standard-library/stdio.h:184:5:184:13 | vasprintf | Reserved identifier 'vasprintf' is declared. |` ... is something in my setup broken?

Reply:
oh, I realized that it is possible to run both 1) `c/cert/src/rules/DCL37-C/DoNotDeclareOrDefineAReservedIdentifier.ql` and 2) `c/common/test/rules/declaredareservedidentifier/DeclaredAReservedIdentifier.ql` in the UI on openpilot, but 2 does not contain the metadata and therefore does not do the filter.  this is what I based including the predicate off of. I guess the `codeql test` is affected in a similar way, still not quite sure what that should actually point to then for a shared query but oh well.

Predicate reverted!

Reply:
also, extra type testcases added and extra query logic also added!

Reply:
@lcartey hm actually, [this is some error](https://github.com/github/codeql-coding-standards/runs/7844753286?check_suite_focus=true) from not using the `exists(d.getFile().getRelativePath())` check I think? or at least I thought so previously, if not, I am not quite sure where that comes from

Reply:
I think the problem in the test case is a bug in the generation of test cases for shared queries.

If we look in the shared library (`DeclaredAReservedIdentifier.qll`) for this query we can see the following parts related to exclusions:
```
abstract class DeclaredAReservedIdentifierSharedQuery extends Query { }

Query getQuery() { result instanceof DeclaredAReservedIdentifierSharedQuery }

query predicate problems(Element m, string message) {
  not isExcluded(m, getQuery()) and
 ....
```
The abstract class here allows any rule query that imports this library to specify the query indirectly. For example, Rule 21.2 has the following:
```
class DoNotDeclareAReservedIdentifierQuery extends DeclaredAReservedIdentifierSharedQuery {
  DoNotDeclareAReservedIdentifierQuery() {
    this = Declarations1Package::doNotDeclareAReservedIdentifierQuery()
  }
}
```

However, the test case we generate for shared rules do not specify any query id:
https://github.com/github/codeql-coding-standards/blob/c05f9a8d30935fbc51ac5e01667d745bd131a618/c/common/test/rules/declaredareservedidentifier/DeclaredAReservedIdentifier.ql#L1-L2

What this means is `isExcluded` will not hold for any source file/query pair, and will therefore not filter results outside the source location.

I'm not certain what the right fix is here, but it's clearly out of the scope of this PR. For now, I think it would be fine to just keep the standard library results in the result set.

Reply:
ok, accepted the library result for this!